### PR TITLE
RTL language support & added arabic localization resources

### LINKED ISSUE #754: Add Arabic Language Resource Translation
Add Arabic language resource translation to all framework modules

### PR DESCRIPTION
**Pull request by kgamalseif**
Added bootstrap-rtl to support RTL at MVC version
Fixes #754

### CODE REVIEW

File: framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs [hunk 1]
```diff
@@ -35,6 +36,10 @@ public static IDisposable Use([NotNull] CultureInfo culture, CultureInfo uiCultu
             });
         }
 
+        public static bool IsRtl
+        {
+            get { return Thread.CurrentThread.CurrentUICulture.TextInfo.IsRightToLeft; }
```
Comment (maliming) [inline framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs:41]: Why not use `CultureInfo.CurrentUICulture`?

Comment (kgamalseif) [reply framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs:41]: Yes, we can but need to be tested.

Actually, I took the code from aspnetzero as it is :)

Comment (maliming) [reply framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs:41]: Because `CultureHelper`  is using `CultureInfo` now. : )

Comment (kgamalseif) [reply framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs:41]: Ok, I will change it and push it to merge.

Comment (kgamalseif) [reply framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs]: Hi @hikalkan @maliming, V2.7 MVC template crashes after upgrade from V2.6

The type or namespace name 'Localization' does not exist in the namespace 'Volo.Abp.AspNetCore' (are you missing an assembly reference?)
+
@using Volo.Abp.AspNetCore.Localization
The name 'CultureHelper' does not exist in the current context
+
<html lang="@CultureInfo.CurrentCulture.Name" dir=@(CultureHelper.IsRtl ? "rtl" : "")>

Also, I got better bootstrap RTL library [https://www.npmjs.com/package/bootstrap-v4-rtl](url)
I made a new style bundle contributor that replaces the bootstrap.css if RTL language.
My question where can I place it at the framework or leave it at the template?

Can I update the pull request if agreed with the new changes?

Thanks,
Khaled

**Comment by kgamalseif** [comment]
Hi @hikalkan @maliming, please check my comment above?

File: framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs (hunk 1 above)
Comment (maliming) [reply framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs]: `CultureHelper` exists in the` Volo.Abp.Core` package, its namespace is `Volo.Abp.Localization`.

Comment (maliming) [reply framework/src/Volo.Abp.Core/Volo/Abp/Localization/CultureHelper.cs]: `style bundle contributor` should be in the framework.

eg: `jstree package & bundle contributor. `https://github.com/abpframework/abp/commit/72b9b43e9130c098a085dbe563a231e8eb6e2f92

**Comment by hikalkan** [comment]
Thanks a lot. I merged. I will test and fix errors if there are.

**Comment by hikalkan** [comment]
image

It seems pretty fine except some minor improvement needs, but we need to handle them.

**Comment by kgamalseif** [comment]
Good, let's test and see how to improve :).

**Comment by kgamalseif** [comment]
I will apply this  https://www.npmjs.com/package/bootstrap-v4-rtl and send another pull request, I tested it and it is very good

**Comment by kgamalseif** [comment]
Hi @maliming ,
could you please help me on the below issue when get latest and try to run mvc app template:

> System.TypeLoadException: Could not load type 'Volo.Abp.Localization.LocalizationResourceNameAttribute' from assembly 'Volo.Abp.Localization, Version=2.8.0.0, Culture=neutral, PublicKeyToken=null'.
>    at System.ModuleHandle.ResolveType(QCallModule module, Int32 typeToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount, ObjectHandleOnStack type)
>    at System.ModuleHandle.ResolveTypeHandleInternal(RuntimeModule module, Int32 typeToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)
>    at System.Reflection.RuntimeModule.ResolveType(Int32 metadataToken, Type[] genericTypeArguments, Type[] genericMethodArguments)
>    at System.Reflection.CustomAttribute.FilterCustomAttributeRecord(MetadataToken caCtorToken, MetadataImport& scope, RuntimeModule decoratedModule, MetadataToken decoratedToken, RuntimeType attributeFilterType, Boolean mustBeInheritable, ListBuilder`1& derivedAttributes, RuntimeType& attributeType, IRuntimeMethodInfo& ctor, Boolean& ctorHasParameters, Boolean& isVarArg)
>    at System.Reflection.CustomAttribute.IsCustomAttributeDefined(RuntimeModule decoratedModule, Int32 decoratedMetadataToken, RuntimeType attributeFilterType, Int32 attributeCtorToken, Boolean mustBeInheritable)
>    at System.Reflection.CustomAttribute.IsDefined(RuntimeType type, RuntimeType caType, Boolean inherit)
>    at System.RuntimeType.IsDefined(Type attributeType, Boolean inherit)
>    at Volo.Abp.DependencyInjection.ConventionalRegistrarBase.IsConventionalRegistrationDisabled(Type type) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\DependencyInjection\ConventionalRegistrarBase.cs:line 38
>    at Volo.Abp.DependencyInjection.DefaultConventionalRegistrar.AddType(IServiceCollection services, Type type) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\DependencyInjection\DefaultConventionalRegistrar.cs:line 16
>    at Volo.Abp.DependencyInjection.ConventionalRegistrarBase.AddTypes(IServiceCollection services, Type[] types) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\DependencyInjection\ConventionalRegistrarBase.cs:line 30
>    at Volo.Abp.DependencyInjection.ConventionalRegistrarBase.AddAssembly(IServiceCollection services, Assembly assembly) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\DependencyInjection\ConventionalRegistrarBase.cs:line 23
>    at Microsoft.Extensions.DependencyInjection.ServiceCollectionConventionalRegistrationExtensions.AddAssembly(IServiceCollection services, Assembly assembly) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Microsoft\Extensions\DependencyInjection\ServiceCollectionConventionalRegistrationExtensions.cs:line 42
>    at Volo.Abp.Modularity.ModuleLoader.ConfigureServices(List`1 modules, IServiceCollection services) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\Modularity\ModuleLoader.cs:line 125
>    at Volo.Abp.Modularity.ModuleLoader.LoadModules(IServiceCollection services, Type startupModuleType, PlugInSourceList plugInSources) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\Modularity\ModuleLoader.cs:line 23
>    at Volo.Abp.AbpApplicationBase.LoadModules(IServiceCollection services, AbpApplicationCreationOptions options) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\AbpApplicationBase.cs:line 80
>    at Volo.Abp.AbpApplicationBase..ctor(Type startupModuleType, IServiceCollection services, Action`1 optionsAction) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\AbpApplicationBase.cs:line 44
>    at Volo.Abp.AbpApplicationWithExternalServiceProvider..ctor(Type startupModuleType, IServiceCollection services, Action`1 optionsAction) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\AbpApplicationWithExternalServiceProvider.cs:line 13
>    at Volo.Abp.AbpApplicationFactory.Create(Type startupModuleType, IServiceCollection services, Action`1 optionsAction) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\AbpApplicationFactory.cs:line 37
>    at Volo.Abp.AbpApplicationFactory.Create[TStartupModule](IServiceCollection services, Action`1 optionsAction) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Volo\Abp\AbpApplicationFactory.cs:line 29
>    at Microsoft.Extensions.DependencyInjection.ServiceCollectionApplicationExtensions.AddApplication[TStartupModule](IServiceCollection services, Action`1 optionsAction) in C:\Users\Khaled\Source\Repos\abp\framework\src\Volo.Abp.Core\Microsoft\Extensions\DependencyInjection\ServiceCollectionApplicationExtensions.cs:line 15
>    at MyCompanyName.MyProjectName.Web.Startup.ConfigureServices(IServiceCollection services) in C:\Users\Khaled\Source\Repos\abp\templates\app\aspnet-core\src\MyCompanyName.MyProjectName.Web\Startup.cs:line 10
>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
>    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
>    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.InvokeCore(Object instance, IServiceCollection services)
>    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass9_0.<Invoke>g__Startup|0(IServiceCollection serviceCollection)
>    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection services)
>    at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass8_0.<Build>b__0(IServiceCollection services)
>    at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services)
>    at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass12_0.<UseStartup>b__0(HostBuilderContext context, IServiceCollection services)
>    at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
>    at Microsoft.Extensions.Hosting.HostBuilder.Build()
>    at MyCompanyName.MyProjectName.Web.Program.Main(String[] args) in C:\Users\Khaled\Source\Repos\abp\templates\app\aspnet-core\src\MyCompanyName.MyProjectName.Web\Program.cs:line 28

**Comment by maliming** [comment]
hi @kgamalseif 

Try run `build-all.ps1`.

https://github.com/abpframework/abp/blob/dev/build/build-all.ps1

**Comment by kgamalseif** [comment]
> I will apply this https://www.npmjs.com/package/bootstrap-v4-rtl and send another pull request, I tested it and it is very good

Sent pull request #4038

**Comment by kgamalseif** [comment]
> hi @kgamalseif
> 
> Try run `build-all.ps1`.
> 
> https://github.com/abpframework/abp/blob/dev/build/build-all.ps1

Hi @maliming , it happens to me again but to project created with CLI after upgrading from 2.7 to 2.8, appreciate your support.

**Comment by maliming** [comment]
This is like a dotnet compilation problem, you can try `dotnet clean dotnet build` etc.

**Comment by Loongle** [comment]
> This is like a dotnet compilation problem, you can try dotnet clean dotnet build etc.
I tried it and it didn't work

**Comment by kgamalseif** [comment]
@Loongle @maliming , did not work for me also, maybe it is related to assembly reference conflicts, kindly see below 3 warnings:
_14>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2081,5): warning MSB3277: Found conflicts between different versions of "Microsoft.Win32.Registry" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
14>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2081,5): warning MSB3277: Found conflicts between different versions of "System.Security.Cryptography.Cng" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed.
14>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2081,5): warning MSB3277: Found conflicts between different versions of "System.Security.Cryptography.Xml" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed._

**Comment by Loongle** [comment]
@kgamalseif I have no warning output. The problem does not seem to be dependency conflict

**Comment by kgamalseif** [comment]
After deleting the packages folder(your profile\.nuget\packages\) and removing unneeded packages from .Web project they were installed before, it works for me :).

**Comment by maliming** [comment]
You can also try to delete the `bin` & `obj` folder of the project

**Pull request merged**